### PR TITLE
Add field visibility functionality

### PR DESCRIPTION
**Pull request by abutterworth**
Doesn't yet save.

### CODE REVIEW

File: src/data/apiClient.js [hunk 1]
```diff
@@ -13,4 +16,69 @@ const apiClient = getAuthenticatedAPIClient({
...
+}
+
+export function savePreferences(username, preferences) {
+  const url = `${lmsBaseUrl}/api/user/v1/preferences/${username}`;
+  // const data = {
+  //   "account_privacy": "custom",
+  //   "visibility.bio": "all_users",
+  //   "visibility.country": "private",
+  //   "visibility.language_proficiencies": "all_users"
+  // };
```
Comment (albemarle) [inline src/data/apiClient.js]: Just checking that leaving this here is intentional.

Comment (abutterworth) [reply src/data/apiClient.js]: at the moment yet

File: src/data/apiClient.js [hunk 2]
```diff
@@ -13,4 +16,69 @@ const apiClient = getAuthenticatedAPIClient({
...
+  education: 'level_of_education',
+  fullName: 'name',
+  username: 'username',
+  profileImage: 'profile_image',
+  dateJoined: 'date_joined',
+  languageProficiencies: 'language_proficiencies',
+  accountPrivacy: 'account_privacy',
+};
+
+const serverClientKeyMap = _.invert(clientServerKeyMap);
```
Comment (abutterworth) [inline src/data/apiClient.js]: A map like this would eliminate our snakekeys and camelcase transformer dependencies and adds clarity.

Comment (albemarle) [reply src/data/apiClient.js]: 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫 🐍 🐫

File: src/data/apiClient.js [hunk 3]
```diff
@@ -13,4 +16,78 @@ const apiClient = getAuthenticatedAPIClient({
...
+  return new Promise((resolve, reject) => {
+    apiClient.get(url)
+      .then(({ data }) => {
+        // Unflatten server response
+        // visibility.social_links: 'value' becomes { visibility: { socialLinks: 'value' }}
+        const preferences = {};
+        Object.entries(data).forEach(([key, value]) => {
+          _.set(
+            preferences,
+            key.split('.').map(pathKey => serverClientKeyMap[pathKey] || pathKey),
```
[the remark this answers is not in the document]
Comment (davidjoy) [reply src/data/apiClient.js]: I realize now that this won’t work, since we don’t know what fields we’re receiving. But... darn it, the old/new versions are still hard to follow. :)

File: src/data/apiClient.js [hunk 4]
```diff
@@ -13,4 +16,78 @@ const apiClient = getAuthenticatedAPIClient({
...
+
+
+export function getPreferences(username) {
+  const url = `${lmsBaseUrl}/api/user/v1/preferences/${username}`;
+
+  return new Promise((resolve, reject) => {
+    apiClient.get(url)
+      .then(({ data }) => {
+        // Unflatten server response
+        // visibility.social_links: 'value' becomes { visibility: { socialLinks: 'value' }}
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply src/data/apiClient.js]: @nasthagiri Is it possible to send objects instead of a flat object with dot syntax keys? From:
```
{
   "visibility.social_links": value
}
```

To:
```
{
   visibility: {
      social_links: value
   }
}
```

[the remark this answers is not in the document]
Comment (nasthagiri) [reply src/data/apiClient.js]: If we created a separate API specifically for, let's say, visibility settings, then we would return it as objects as proposed.  However, we're currently re-using the existing "user preferences" API since that's where visibility preferences have been historically accessed.  So we have extended that API to store account-level visibility preferences.

Keep in mind that "visibility settings" for other "pluggable features", like Course Certificates, would come from a different Course-Certs-specific API.

[the remark this answers is not in the document]
Comment (nasthagiri) [reply src/data/apiClient.js]: We can discuss this further.

File: package.json [hunk 5]
```diff
@@ -44,6 +44,7 @@
     "font-awesome": "^4.7.0",
     "history": "^4.7.2",
     "i18n-iso-countries": "^3.7.8",
+    "lodash": "^4.17.11",
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply package.json]: 👍

File: src/components/UserProfile/index.jsx [hunk 6]
```diff
@@ -44,11 +44,25 @@ class UserProfile extends React.Component {
...
-  onSave(fieldName, value) {
+  onSave(fieldName, value, visibility) {
+    const {
+      value: fieldValue,
+      visibility: fieldVisibility,
+    } = this.state[fieldName];
+
     const userAccountData = {
-      [fieldName]: value || this.state[fieldName].value,
+      [fieldName]: value || fieldValue,
```
[the remark this answers is not in the document]
Comment (abutterworth) [reply src/components/UserProfile/index.jsx]: We spoke in person, but for posterity: components that need to manage their own dirty state for whatever reason (in our current case social links) can provide the value directly to onSave rather than saving what's in the parent component state.

File: src/data/apiClient.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (abutterworth) [reply src/data/apiClient.js]: I broke these out into utils files, so it should be easier to follow. Not to say this is resolved, but for now at least it's easier to understand at first glance.
